ledger/catchpointtracker.go

```diff
@@ -399,57 +419,330 @@ func (ct *catchpointTracker) commitRound(ctx context.Context, tx *sql.Tx, dcc *d
+		dcc.stats.MemoryUpdatesDuration = time.Duration(time.Now().UnixNano())
+	}
 }
 
-func (ct *catchpointTracker) postCommit(ctx context.Context, dcc *deferredCommitContext) {
-	var err error
-	if dcc.isCatchpointRound {
-		dcc.catchpointLabel, err = ct.accountsCreateCatchpointLabel(dcc.newBase+dcc.lookback, dcc.roundTotals, dcc.committedRoundDigest, dcc.trieBalancesHash)
+func doRepackCatchpoint(header CatchpointFileHeader, in *tar.Reader, out *tar.Writer) error {
+	{
+		bytes := protocol.Encode(&header)
+
+		err := out.WriteHeader(&tar.Header{
+			Name: "content.msgpack",
+			Mode: 0600,
+			Size: int64(len(bytes)),
+		})
 		if err != nil {
-			ct.log.Warnf("commitRound : unable to create a catchpoint label: %v", err)
+			return err
+		}
+
+		_, err = out.Write(bytes)
+		if err != nil {
+			return err
 		}
 	}
-	if ct.balancesTrie != nil {
-		_, err = ct.balancesTrie.Evict(false)
+
+	for {
+		header, err := in.Next()
 		if err != nil {
-			ct.log.Warnf("merkle trie failed to evict: %v", err)
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+
+		buf := make([]byte, header.Size)
+		bytesRead := int64(0)
+		for (err == nil) && (bytesRead < header.Size) {
+			var x int
+			x, err = in.Read(buf[bytesRead:])
+			bytesRead += int64(x)
+		}
+		if (err != nil) && (err != io.EOF) {
+			return err
+		}
+
+		err = out.WriteHeader(header)
+		if err != nil {
+			return err
+		}
+
+		_, err = out.Write(buf)
+		if err != nil {
+			return err
 		}
 	}
+}
 
-	ct.catchpointsMu.Lock()
-	ct.roundDigest = ct.roundDigest[dcc.offset:]
-	if dcc.isCatchpointRound && dcc.catchpointLabel != "" {
-		ct.lastCatchpointLabel = dcc.catchpointLabel
+func repackCatchpoint(header CatchpointFileHeader, dataPath string, outPath string) error {
+	// Initialize streams.
+	fin, err := os.OpenFile(dataPath, os.O_RDONLY, 0666)
+	if err != nil {
+		return err
+	}
+	defer fin.Close()
+
+	gzipIn, err := gzip.NewReader(fin)
+	if err != nil {
+		return err
 	}
+	defer gzipIn.Close()
+
+	tarIn := tar.NewReader(gzipIn)
+
+	fout, err := os.OpenFile(outPath, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return err
+	}
+	defer fout.Close()
+
+	gzipOut := gzip.NewWriter(fout)
+	defer gzipOut.Close()
+
+	tarOut := tar.NewWriter(gzipOut)
+	defer tarOut.Close()
+
+	// Repack.
+	err = doRepackCatchpoint(header, tarIn, tarOut)
+	if err != nil {
+		return err
+	}
+
+	// Close streams.
+	err = tarOut.Close()
+	if err != nil {
+		return err
+	}
+
+	err = gzipOut.Close()
+	if err != nil {
+		return err
+	}
+
+	err = fout.Close()
+	if err != nil {
+		return err
+	}
+
+	err = gzipIn.Close()
+	if err != nil {
+		return err
+	}
+
+	err = fin.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Create a catchpoint (a label and possibly a file with db record).
+func (ct *catchpointTracker) createCatchpoint(accountsRound basics.Round, round basics.Round, dataInfo catchpointFirstStageInfo, blockHash crypto.Digest) error {
+	label := ledgercore.MakeCatchpointLabel(
```

Comment:
`MakeCatchpointLabel` is called only here but the function does not quit if only label is required